### ISSUE TITLE
Can an extra output file be added as a checkpoint for the Snakemake workflow?

### ISSUE THREAD

**Issue by evolgen** [iRNA-COSI/APAeval#46]
Hi developers,

I was wondering if it makes sense to add an extra output file for each rule as a checkpoint, something in the lines of 
```
output:        
    out=os.path.join(config["out_dir"], "{sample}", "execute.out"),
    finisher=os.path.join(config["out_dir"], "{sample}", "execute.done")
shell:        
    "(EXECUTE_COMMAND \            
        -i {input.bam} \            
        -o {output.out} \            
        -p1 {params.param1} \            
        -p2 {params.param2}) \            &> {log};
        echo "DONE" >{output.finisher}"
```

This acts as a checkpoint for each rule, in cases where the program(s) partially executes or sometimes the node exits abruptly.
Thank you.

**Comment by mfansler** [comment]
Thanks for the suggestion!

I'm not opposed to adding such checkpoint flags in principle, but practically it's not clear to me that it catches anything Snakemake wouldn't already handle. Do you know of specific experiences (say, with version 5+) where it failed without them?

Perhaps this is different for AWS execution, but my experience with Snakemake on HPCs (SLURM, LSF) has been that it consistently uses `set -e` on submitted job scripts, propagating and catching errors. For example, I've had cases where a typo in the very last command (e.g., `rm my_tmp_dir` without a `-r`) triggered an error, leading Snakemake to mark the rule execution invalid and delete all expected outputs. Mysteriously vanishing jobs have led it to wait indefinitely (at least on LSF), and after interrupting and restarting, it used residual lock files it used to detect that the job never signalled completion. So for me, I've found it sufficiently defensive by default.

**Comment by evolgen** [comment]
I have had an experience (not differential expression, rather genome annotation) where the process skipped the last few chromosomes, but the process didn't produce an error. I faced the same when running a tool suite which aggregates multiple software to produce a combined result (summarizing SNP calls). I need to add that this was on a slurm batch submission job sessions. This was a practice I added in cases where some software do not necessarily mention if the job finished successfully, or for scenarios where there were too many log files to check.

I fully agree that the folder lock works exceptionally well. This is just a suggestion, please feel free to ignore if not required :)

**Comment by mfansler** [comment]
Ah makes sense - good to know when to look out for this. It certainly doesn't hurt to add. Perhaps a `touch {output.finisher}` instead of echoing and make it `.execute.done` to keep it out of plain view?

**Comment by evolgen** [comment]
Yep, touch is definitely much better, echo was my habit of making a git commit messages 👍
